### PR TITLE
fix: (aws-ec2): fix vpc endpoint incorrect issue in China region

### PR DESCRIPTION
**Pull request by readybuilderone**
fix:(aws-ec2): fix vpc endpoint incorrect issue in China region

This PR fix the issue that can't create interface vpc endpoint for 40+ services like ecr, ec2, athena etc

Closes: #9864

----

# Considerations 
1. In cn-north-1 region, there are:
  - 2 services whose endpoint service begin with "aws.sagemaker" prefix;
  - 41 services whose endpoint service begin with "cn.com.amazonaws" prefix;
  - 21 services whose endpoint service begin with "com.amazonaws" prefix;
    Details: https://gist.github.com/readybuilderone/d355f9f8f0f2b66379a10742b3c67cc7
2. in cn-northwest-1 region, there are:
  - 2 services whose endpoint service begin with "aws.sagemaker" prefix;
  - 44 services whose endpoint service begin with "cn.com.amazonaws" prefix;
  - 21 services whose endpoint service begin with "com.amazonaws" prefix;
    Details: https://gist.github.com/readybuilderone/a79f2c5e6fa02aae1699bf674b08be7c

So, In cn-north-1 and cn-northwest-1, the vpc endpoint prefix is both region and service related.

At first, I found the vpc endpoint prefix could be fetched via [AWS API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeVpcEndpoints.html) , however  the javascript sdk supports only asynchronous call, currently I didn't find any pattern in CDK to make live calls in object construction.  Since the exception rules won't ever change ( only be added to), and the rate of change should be small in the grand scheme, I just kept the exception prefix lists  in a map as fact data. 

# Known Issues
Since the interface vpc endpoints prefixs are not region agnostic, it requires to set the {region} in the environment to create stacks in cn-north-1, and cn-northwest-1. The methods could be found in the [offical guide](https://docs.aws.amazon.com/cdk/latest/guide/environments.html).

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (pahud) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:329]:
```suggestion
    const defaultEndpointPrefix =Lazy.uncachedString({
```
How about just call it `defaultEndpointPrefix` ?

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (pahud) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:347]: Let's wrap the lines to make it more readable.

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (pahud) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:343]: Maybe we can add an AWS CLI sample in the doc comments showing how to get all endpoints from a region. That would be helpful.

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (pahud) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:349]:
```suggestion
    if (VPC_ENDPOINT_SERVICE_EXCEPTIONS[region] && VPC_ENDPOINT_SERVICE_EXCEPTIONS[region].includes(name)) {
```

I believe this should be working.

**Review by pahud — CHANGES_REQUESTED** [review]
just some minor suggested changes.

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (readybuilderone) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:329]: Thanks @Pahud, that's a good suggestion, I'll update it.

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (readybuilderone) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:349]: It will be more concise, thanks Pahud for your suggestion, I'll update it

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (readybuilderone) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:343]: ok, I'll add the AWS CLI sample to the comments

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (readybuilderone) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:347]: good suggestions, thanks Pahud

**Comment by readybuilderone** [comment]
Thanks @Pahud for review my PR, I have updated my code according to your suggestions.
@njlynch can you help to assign a SDE to review my PR? Thanks.

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:329]:
```suggestion
    const defaultEndpointPrefix = Lazy.uncachedString({
```

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:363]:
```suggestion
    if (VPC_ENDPOINT_SERVICE_EXCEPTIONS[region]?.includes(name)) {
```

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:354]: `!` I think `transcribe.cn` is an _additional_ edge case. Based on the data here, it looks like it's the single case of a region-specific suffix. As-is, the `name` input to this function will never be `transcribe.cn`, so the exception will never be properly picked up. I guess we could create a corresponding getDefaultEndpointSuffix with just transcribe as the exception? What do you think?

**Review by njlynch — DISMISSED** [review]
Thanks @readybuilderone , and thanks @pahud  for the first-round review!

I think it's worth getting the `transcribe` edge case right before we merge this; otherwise it looks great.

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (readybuilderone) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:354]: You're right, the .cn suffix should be handled for transcribe in China region.

``` typescript
this.name = `${prefix || defaultEndpointPrefix}.${region}.${name}${defaultEndpointSuffix}`;
```

I'll implement the getDefaultEndpointSuffix function and add the relative unit tests.

Comment (readybuilderone) [reply packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts]: I have updated my implementation and added the unit tests. Am I doing the right way? @njlynch

File: packages/@aws-cdk/aws-ec2/test/vpc-endpoint.test.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/test/vpc-endpoint.test.ts:678]:
```suggestion
    'test vpc interface endpoint for transcribe can be created correctly in non-china regions'(test: Test) {
```

File: packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts:395]:
```suggestion
    return VPC_ENDPOINT_SERVICE_EXCEPTIONS[region]?.includes(name) ? '.cn' : '';
```

**Review by njlynch — APPROVED** [review]
Looks great, thanks!

**Comment by pahud** [comment]
Looks like there is a `validate-pr` waiting for approval for first-time contributor and probably need @njlynch kick it off.

**Comment by readybuilderone** [comment]
> Looks like there is a `validate-pr` waiting for approval for first-time contributor and probably need @njlynch kick it off.

Yes, this is my first PR to CDK, thanks @pahud  and @njlynch so much!
I'm not quite clear about the process, so just want to confirm with @njlynch , is there anything I should do before the approval?

**Review by njlynch — APPROVED** [review]

**Comment by njlynch** [comment]
> is there anything I should do before the approval?

Nope, you're good. I just had to re-click the approve button, as the review was dismissed after the merge. It's in the queue to be automatically merged now, and should be merged shortly.

**Pull request merged**
